### PR TITLE
feat/conformance 004

### PR DESCRIPTION
**Pull request by Mirko-von-Leipzig**
I took a stab at conformance test case 004. I think I've managed to come up with a strategy that neatly tests all the cases without arbitrary sleeps etc.

### CODE REVIEW

File: src/tests/conformance/handshake.rs [hunk 1]
```diff
@@ -72,3 +75,90 @@ async fn handshake_initiator_side() {
 
     node.stop().await;
 }
+
+#[tokio::test]
+async fn conformance_004() {
```
Comment (Mirko-von-Leipzig) [inline src/tests/conformance/handshake.rs:80]: I'm not sure what to name these tests. I'd prefer names that indicate what they're testing, but some prefer being able to trace the spec to the test.

File: src/tests/conformance/handshake.rs [hunk 2]
```diff
@@ -72,3 +75,90 @@ async fn handshake_initiator_side() {
...
+    // The node should reject non-Version messages in response to the initial Version it sent.
```
Comment (niklaslong) [inline src/tests/conformance/handshake.rs:81]: Thanks for the great description here! 🙌

**Review by niklaslong — COMMENTED** [review]
Great work on this! 🎉 
Left a few comments/suggestions.

File: src/tests/conformance/handshake.rs [hunk 3]
```diff
@@ -72,3 +75,90 @@ async fn handshake_initiator_side() {
...
+    let (zig, node_meta) = read_config_file();
+
+    for bad_message in bad_messages {
+        let listener = TcpListener::bind(zig.new_local_addr()).await.unwrap();
+
+        let mut node = Node::new(node_meta.clone());
+        node.initial_peers(vec![listener.local_addr().unwrap().port()]);
+        node.start().await;
+
+        match listener.accept().await {
```
[the remark this answers is not in the document]
Comment (Mirko-von-Leipzig) [reply src/tests/conformance/handshake.rs:133]: Oh right - I'd just copied this from the handshake test.

File: src/tests/conformance/handshake.rs (hunk 1 above)
Comment (niklaslong) [reply src/tests/conformance/handshake.rs:80]: How about naming them nicely whilst commenting the spec ID just under it? At the very least, this should make it easy for us to switch later if necessary.

File: src/tests/conformance/handshake.rs [hunk 4]
```diff
@@ -72,3 +75,90 @@ async fn handshake_initiator_side() {
...
+        // Message::Inv(Inv));
+        // Message::NotFound(Inv));
+    ];
+
+    let (zig, node_meta) = read_config_file();
+
+    for bad_message in bad_messages {
+        let listener = TcpListener::bind(zig.new_local_addr()).await.unwrap();
+
+        let mut node = Node::new(node_meta.clone());
```
[the remark this answers is not in the document]
Comment (Mirko-von-Leipzig) [reply src/tests/conformance/handshake.rs:129]: The issue I ran into was that the zcashd node doesn't initiate a handshake automatically then (I may simply not be waiting long enough?).

Can we instruct the node to add a peer via RPC?

File: src/tests/conformance/handshake.rs [hunk 5]
```diff
@@ -72,3 +75,90 @@ async fn handshake_initiator_side() {
...
+                // (3)
+                match Message::Version(Version::new(addr, listener.local_addr().unwrap()))
+                    .write_to_stream(&mut stream)
+                    .await
+                {
+                    Ok(_) => {
+                        // (4)
+                        match Message::read_from_stream(&mut stream).await {
+                            Ok(message) => assert!(matches!(message, Message::Verack)),
+                            Err(ref err) if err.kind() == std::io::ErrorKind::UnexpectedEof => {}
```
[the remark this answers is not in the document]
Comment (Mirko-von-Leipzig) [reply src/tests/conformance/handshake.rs:151]: Sure, makes sense to me.

**Comment by Mirko-von-Leipzig** [comment]
I've updated the PR to use multiple "peer" nodes concurrently instead of restarting the node for every message.

**Review by niklaslong — APPROVED** [review]
LGTM 👍

**Pull request merged**